src/stacked_borrows.rs

```diff
@@ -420,43 +432,39 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for super::MiriEvalContext<'a, 'mir, '
         pointee_ty: Ty<'tcx>,
         size: Size,
-        ref_kind: RefKind,
+        usage: UsageKind,
     ) -> EvalResult<'tcx, Borrow> {
+        trace!("tag_reference: Accessing reference ({:?}) for {:?} (pointee {}, size {})",
+            usage, ptr, pointee_ty, size.bytes());
         // In principle we should not have to do anything here.  However, with transmutes involved,
-        // it can happen that the tag of `ptr` does not actually match `ref_kind`, and we
+        // it can happen that the tag of `ptr` does not actually match `usage`, and we
         // should adjust for that.
         // Notably, the compiler can introduce such transmutes by optimizing away `&[mut]*`.
         // That can transmute a raw ptr to a (shared/mut) ref, and a mut ref to a shared one.
-        match (ref_kind, ptr.tag) {
-            (RefKind::Raw, _) => {
+        match (usage, ptr.tag) {
+            (UsageKind::Raw, _) => {
                 // Don't use the tag, this is a raw access!  Even if there is a tag,
                 // that means transmute happened and we ignore the tag.
                 // Also don't do any further validation, this is raw after all.
                 return Ok(Borrow::Mut(Mut::Raw));
             }
-            (RefKind::Mut, Borrow::Mut(Mut::Uniq(_))) |
-            (RefKind::Shr, Borrow::Frz(_)) |
-            (RefKind::Shr, Borrow::Mut(Mut::Raw)) => {
+            (UsageKind::Write, Borrow::Mut(Mut::Uniq(_))) |
+            (UsageKind::Read, Borrow::Frz(_)) |
+            (UsageKind::Read, Borrow::Mut(Mut::Raw)) => {
                 // Expected combinations.  Nothing to do.
                 // FIXME: We probably shouldn't accept this if we got a raw shr without
                 // interior mutability.
             }
-            (RefKind::Mut, Borrow::Mut(Mut::Raw)) => {
+            (UsageKind::Write, Borrow::Mut(Mut::Raw)) => {
                 // Raw transmuted to mut ref.  Keep this as raw access.
                 // We cannot reborrow here; there might be a raw in `&(*var).1` where
                 // `var` is an `&mut`.  The other field of the struct might be already frozen,
                 // also using `var`, and that would be okay.
             }
-            (RefKind::Shr, Borrow::Mut(Mut::Uniq(_))) => {
-                // A mut got transmuted to shr.  High time we freeze this location!
-                // Make this a delayed reborrow.  Redundant reborows to shr are okay,
-                // so we do not have to be worried about doing too much.
-                // FIXME: Reconsider if we really want to mutate things while doing just a deref,
-                // which, in particular, validation does.
-                trace!("tag_dereference: Lazy freezing of {:?}", ptr);
-                return self.tag_reference(ptr, pointee_ty, size, ref_kind);
+            (UsageKind::Read, Borrow::Mut(Mut::Uniq(_))) => {
+                // A mut got transmuted to shr.  The mut borrow must be reactivatable.
```

Comment:
can you add a code example where this would happen? This case is very.. unusual

Reply:
```rust
let x = &mut 3;
let y = &*x; // compiler optimizes to `let y = x;`, effectively a transmute
```